src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs

```diff
@@ -0,0 +1,169 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+namespace Peachpie.CodeAnalysis.Utilities
+{
+    /// <summary>
+    /// PriorityQueue provides a stack-like interface, except that objects
+    /// "pushed" in arbitrary order are "popped" in order of priority, i.e., 
+    /// from least to greatest as defined by the specified comparer.
+    /// </summary>
+    /// <remarks>
+    /// Push and Pop are each O(log N). Pushing N objects and them popping
+    /// them all is equivalent to performing a heap sort and is O(N log N).
+    /// Multiple different items of the same priority are allowed, but their
+    /// order is not guaranteed to be stable.
+    /// 
+    /// The original implementation was taken from the Microsoft .NET
+    /// Framework Reference Source, author: Niklas Borson (niklasb).
+    /// </remarks>
+    internal class PriorityQueue<T>
+    {
+        private T[] _heap;
+        private int _count;
+        private IComparer<T> _comparer;
+        private const int DefaultCapacity = 6;
+
+        public PriorityQueue(IComparer<T> comparer)
+        {
+            _heap = new T[DefaultCapacity];
+            _count = 0;
+            _comparer = comparer;
+        }
+
+        /// <summary>
+        /// Gets the number of items in the priority queue.
+        /// </summary>
+        public int Count
+        {
+            get { return _count; }
+        }
+
+        /// <summary>
+        /// Gets the first or topmost object in the priority queue, which is the
+        /// object with the minimum value.
+        /// </summary>
+        public T Top
+        {
+            get
+            {
+                Debug.Assert(_count > 0);
+                return _heap[0];
+            }
+        }
+
+        /// <summary>
+        /// Adds an object to the priority queue.
+        /// </summary>
+        internal void Push(T value)
```

Comment:
is it thread safe? always called within lock ?

Reply:
No, ```PriorityQueue``` is a simple heap with no locking. Thread safety is provided by ```DistinctQueue```.